### PR TITLE
refactor!(apis): add `WALProvider` and `FileStorage`

### PR DESCRIPTION
**Pull request by zyy17**
### APIs Breaking

1. Add `WALProvider` to support `RaftEngineWAL` and `KafkaWAL`;
2. Add `FileStorage` to support creating standalone wal or cache in different file system storage;

## Enhancement

1. Update the code comment of APIs;
2. Add multiple field accessors of `GreptimeDBCluster` and `GreptimeDBStandalone`;

## Summary by CodeRabbit

- **New Features**
	- Enhanced API definitions for GreptimeDB with expanded phase and component specifications.
	- Introduced new storage options, including detailed configurations for file and object storage.
	- Added support for Write-Ahead Logging (WAL) with both Raft and Kafka options.
	- Introduced a new configuration for standalone GreptimeDB instances with comprehensive specifications.
	- Added a new end-to-end test for GreptimeDB cluster with standalone WAL configuration.

- **Bug Fixes**
	- Improved validation logic for configuration parameters, ensuring required fields are enforced.

- **Documentation**
	- Updated API references with detailed descriptions of new and modified structures.
	- Enhanced documentation for configuration schemas and requirements.
	- Added examples for standalone configurations to enhance user guidance.

- **Chores**
	- Refactored code for better clarity and maintainability, including the consolidation of constants and utility functions.

### CODE REVIEW

File: apis/v1alpha1/common.go [hunk 1]
```diff
@@ -183,68 +193,68 @@ type MainContainerSpec struct {
...
 	// Image pull policy.
-	// One of Always, Never, IfNotPresent.
-	// Defaults to Always if :latest tag is specified, or IfNotPresent otherwise.
+	// One of `Always`, `Never`, `IfNotPresent`.
+	// Defaults to `Always` if `:latest` tag is specified, or IfNotPresent otherwise.
 	// Cannot be updated.
-	// More info: https://kubernetes.io/docs/concepts/containers/images#updating-images
-	// ImagePullPolicy field is from 'corev1.Container.ImagePullPolicy'.
+	// More info: `https://kubernetes.io/docs/concepts/containers/images#updating-images`
+	// ImagePullPolicy field is from `orev1.Container.ImagePullPolicy`.
```
Comment (daviderli614) [inline apis/v1alpha1/common.go:257]: Is `corev1` need to add `c` prefix.

File: apis/v1alpha1/common.go [hunk 2]
```diff
@@ -413,31 +619,37 @@ type GCSStorageProvider struct {
...
+	// Enabled indicates whether the PodMonitor is enabled.
+	// +required
+	Enabled bool `json:"enabled"`
 
-	// Prometheus PodMonitor labels.
-	// +optional
-	Labels map[string]string `json:"labels,omitempty"`
+	// Labels is the labels for the PodMonitor.
+	// +required
+	Labels map[string]string `json:"labels"`
```
Comment (daviderli614) [inline apis/v1alpha1/common.go:639]: Suggestion: Maybe is optional, not required field.

File: apis/v1alpha1/greptimedbcluster_types.go [hunk 3]
```diff
@@ -18,11 +18,11 @@ import (
...
 )
 
-// ComponentSpec is the common specification for all components(frontend/meta/datanode).
+// ComponentSpec is the common specification for all components(`frontend`/`meta`/`datanode`/`flownode`).
 type ComponentSpec struct {
 	// The number of replicas of the components.
-	// +optional
-	// +kubebuilder:validation:Minimum=0
+	// +required
+	// +kubebuilder:validation:Minimum=1
```
Comment (daviderli614) [inline apis/v1alpha1/greptimedbcluster_types.go:25]: I think the replicas minimum is 0, and sometimes it is necessary to set the replicas count to 0 to disconnect the frontend network connection.

**Comment by daviderli614** [comment]
What difference from the [Spec.Datanode.Storage](https://github.com/GreptimeTeam/greptimedb-operator/pull/180/files#diff-dde5059d6a7c85d4f5225aeae354403b15aad079989d1d3462ae8af53eb416cdR143) and [Spec.DatanodeStorage](https://github.com/GreptimeTeam/greptimedb-operator/pull/180/files#diff-e40df7f2087abb7f194175cc42122466d50ddd71aeeb694a7de5491b8d4ab03aR69) ?

**Comment by zyy17** [comment]
> What difference from the [Spec.Datanode.Storage](https://github.com/GreptimeTeam/greptimedb-operator/blob/d207e66dad42dfa7e2945e888b87f6dfb9982565/apis/v1alpha1/greptimedbcluster_types.go#L143) and [Spec.DatanodeStorage](https://github.com/GreptimeTeam/greptimedb-operator/blob/d207e66dad42dfa7e2945e888b87f6dfb9982565/apis/v1alpha1/greptimedbstandalone_types.go#L69) ? Maybe same naming ?

It's the same thing. I think their name is almost the same.

**Comment by daviderli614** [comment]
Can we add a testing of mounting `wal` and `cache` on different disks?

**Review by daviderli614 — APPROVED** [review]
LGTM

**Pull request merged**